### ISSUE TITLE
CH3 Keras和TensorFlow簡介

### ISSUE THREAD

**Issue by Willie42825409** [WanXuanShenTW/NTUBIMD114402#10]
## 第三章的各個程式解釋與實際操作
```python
# %%
import tensorflow as tf
from tensorflow import keras
from keras import models
from keras import layers
import numpy as np
import matplotlib.pyplot as plt

#3-5-1 常數張量與變數
# %% 3.1 創建完全由1或0組成的張量
#建立元素值皆為1的張量
x = tf.ones(shape=(2, 1))
print(x)
#建立元素值皆為0的張量
x = tf.zeros(shape=(2, 1))
print(x) 

# %% 3.2 創建由亂數組成的張量
#依據常態分佈(mean=0, stddev=1)建立n*m個元素值並儲存於張量
n, m = 3, 1
x = tf.random.normal(shape=(n, m), mean=0, stddev=1)
print(x)

#在範圍內(max=1, min=0)隨機生成n*1個元素值並儲存於張量
n, m = 3, 1
x = tf.random.uniform(shape=(n, m), minval=0, maxval=1)
print(x)

#書中的抽取的意思是指由指定的生成邏輯隨機生成並儲存 並不是生成完一整個資料後再隨機抽取

# %% 3.3 NumPy 陣列可接受指派新的值
#可以更改NumPy陣列中的元素值
x = np.ones((2, 2))
print(x)
x[0, 0] = 0
print(x)

# %% 3.4 TensorFlow 張量不可接受指派新的值
#TensorFlow張量無法更改元素值
x = tf.ones((2, 2))
print(x)
# %%
x[0, 0] = 0 #會產生錯誤
print(x)

#%% 3.5 創建 Variable 物件
#依據常態分佈(預設 mean=0, stddev=1)建立元素值並儲存於3*1的Variable物件
v = tf.Variable(initial_value=tf.random.normal(shape=(3, 1))) #依據常態分佈(預設 mean=0, stddev=1)建立元素值並儲存
print(v)

#%% 3.6 為 Variable 物件指派一個值
v = tf.Variable(initial_value=tf.random.normal(shape=(3, 1))) #依據常態分佈(預設 mean=0, stddev=1)建立元素值並儲存
v.assign(tf.ones((3, 1))) #指派3*1全部元素值為1
print(v)

#%% 3.7 對 Variable 物件進行局部修改
v = tf.Variable(initial_value=tf.random.normal(shape=(3, 1))) #依據常態分佈(預設 mean=0, stddev=1)建立元素值並儲存
print(v)
v[0, 0].assign(3) #指派第0列第0行元素值為3
print(v)

#%% 3.8 使用 assign_add() 函式
v = tf.Variable(initial_value=tf.random.normal(shape=(3, 1))) #依據常態分佈(預設 mean=0, stddev=1)建立元素值並儲存
print(v)
v.assign_add(tf.ones((3, 1))) #將3*1全部元素值加1
print(v)

#3-5-2 張量操作 : 在TensorFlow中進行數學運算
#%% 3.9 一些基本的數學操作
a = tf.ones((2, 2)) #建立2*2元素值皆為1的張量
a *= 2 #將變為自身的兩倍
print(a)
b = tf.square(a)    #對a中的每個元素平方
print(b)
c = tf.sqrt(a)  #對a中的每個元素開根號
print(c)
d = b + c   #將b和c中的元素進行相加
print(d)
e = tf.matmul(a, b) #將a和b進行矩陣相乘
print(e)
e *= d  #將e中的每個元素分別乘以d中的對應元素
print(e)

#3-5-3 GradientTape API 的進一步說明
#%% 3.10 使用 GradientTape
input_var = tf.Variable(initial_value=3.) #建立一個Variable物件, 初始值為3
with tf.GradientTape() as tape: #建立一個GradientTape物件(在這個範圍內會持續記錄梯度)
    result = tf.square(input_var) #對input_var中的元素進行平方
gradient = tape.gradient(result, input_var) #計算result對input_var的梯度(變化量)
print(gradient) #輸出梯度值(變化量)

#%% 3.11 在 GradientTape 中使用常數張量為輸入
input_const = tf.constant(3.) #建立一個常數張量, 值為3
print(input_const)
with tf.GradientTape() as tape: #建立一個GradientTape物件(在這個範圍內會持續記錄梯度)
    tape.watch(input_const) #將input_const加入梯度計算列表, 因為tf.constant()預設不會被追蹤 
    result = tf.square(input_const) #對input_const中的元素進行平方
gradient = tape.gradient(result, input_const) #計算result對input_const的梯度(變化量)
print(gradient) #輸出梯度值(變化量)

#%% 3.12 利用巢狀梯度磁帶計算二階梯度
time = tf.Variable(1.) #建立一個Variable物件, 初始值為1
with tf.GradientTape() as outer_tape: #建立一個GradientTape物件(在這個範圍內會持續記錄梯度)
    with tf.GradientTape() as inner_tape: #建立一個GradientTape物件(在這個範圍內會持續記錄梯度)
        position = 4.9 * (time ** 2) #計算位置
    speed = inner_tape.gradient(position, time) #計算速度
acceleration = outer_tape.gradient(speed, time) #計算加速度
print(speed) #輸出速度值(變化量)
print(acceleration) #輸出加速度值(變化量)

# 3-5-4 端到端的範例 : 使用 TensorFlow 建立線性分類器
#%%3.13 在平面上生成兩個類別的隨機資料點 (到3.21都是一體的)
num_samples_per_class = 1000 #每個類別的資料點數
negative_samples = np.random.multivariate_normal(
    mean=[0, 3], #第一類別資料點的中心
    cov=[[1, 0.5],[0.5, 1]], #第一類別資料點的限制範圍
    size=num_samples_per_class) #生成第一類別的資料點
positive_samples = np.random.multivariate_normal(
    mean=[3, 0], #第二類別資料點的中心
    cov=[[1, 0.5],[0.5, 1]], #第二類別資料點的限制範圍
    size=num_samples_per_class) #生成第二類別的資料點

#%% 3.14 將資料點堆疊成 shape 為(2000,2)的新陣列
inputs = np.vstack((negative_samples, positive_samples)).astype(np.float32) #將兩類別的資料點堆疊成新陣列
#%% 3.15 生成對應的目標值(0或1)
targets = np.vstack((np.zeros((num_samples_per_class, 1), dtype='float32'), #生成第一類別的目標值
                     np.ones((num_samples_per_class, 1), dtype='float32'))) 
#%% 3.16 繪製兩個類別的資料點
plt.scatter(inputs[:, 0], inputs[:, 1], c=targets[:, 0]) #繪製資料點 X座標:inputs[:, 0], Y座標:inputs[:, 1], 目標值:targets[:, 0]
plt.show() #顯示圖表

#%% 3.17 創建分類器中的變數
input_dim = 2 #輸入維度(資料點的維度)
output_dim = 1 #輸出維度(樣本預測分數的維度)
W = tf.Variable(initial_value=tf.random.uniform(shape=(input_dim,output_dim)))
b = tf.Variable(initial_value=tf.zeros(shape=(output_dim,)))
#%% 3.18 正向傳播函式
def model(inputs):
    return tf.matmul(inputs, W) + b #計算線性分類器的輸出
#%% 3.19 均方誤差損失函數
def square_loss(targets, predictions):
    per_sample_loss = tf.square(targets - predictions) #計算每個樣本的損失
    return tf.reduce_mean(per_sample_loss) #計算所有樣本的平均損失
#%% 3.20 訓練函式
learning_rate = 0.1 #學習率
def training_step(inputs, targets):
    with tf.GradientTape() as tape: #建立一個GradientTape物件(在這個範圍內會持續記錄梯度)
        predictions = model(inputs) #計算模型的輸出
        loss = square_loss(targets, predictions) #計算模型的損失
    grad_loss_wrt_W, grad_loss_wrt_b = tape.gradient(loss, [W, b]) #計算損失對W和b的梯度
    
    W.assign_sub(grad_loss_wrt_W * learning_rate) #損失對W的梯度(差值)乘以學習率後更新W
    b.assign_sub(grad_loss_wrt_b * learning_rate) #損失對b的梯度(差值)乘以學習率後更新b
    return loss
#深度學習的核心程式，裡面有計算損失與自動依據學習率與梯度調整權重與偏移量的功能
#%% 3.21 批次訓練迴圈
for step in range(50):
    loss = training_step(inputs, targets) #進行一次訓練
    print('Loss at step %d: %.4f' % (step, loss)) #輸出損失值
#增加分隔線
z = 0.5 #分隔線的閾值(兩邊以多少比例分隔)
x = np.linspace(-1, 4, 100) #生成-1到4之間的100個數字
y = -W[0] / W[1] * x + (z - b) / W[1] #利用最後優化完成的權重計算分隔線的y座標
#-W[0]/W[1]是斜率，(z-b)/W[1]是截距
plt.plot(x, y, '-r') #繪製分隔線
# 繪製輸出結果
predictions = model(inputs) #計算模型的輸出
plt.scatter(inputs[:, 0], inputs[:, 1], c=predictions[:, 0] > z) #繪製資料點
plt.show() #顯示圖表

#3-6-1 Layer (層) : 深度學習的基石
#%% 3.22 使用 Layer 類別來實作 Dense 層
class SimpleDense(keras.layers.Layer): #繼承Layer類別
    def __init__(self, units, activation=None):
        super().__init__()
        self.units = units 
        self.activation = keras.activations.get(activation) 
    def build(self, input_shape): #建立權重
        input_dim = input_shape[-1] #取得輸入維度
        self.W = self.add_weight(shape=(input_dim, self.units), initializer='random_normal') #建立權重W
        self.b = self.add_weight(shape=(self.units,), initializer='zeros') #建立偏移量b
    def call(self, inputs): #定義正向傳播
        y = tf.matmul(inputs, self.W) + self.b #計算線性組合
        if self.activation is not None: #如果有設定激活函數
            y = self.activation(y) #套用激活函數
        return y #回傳輸出
my_dense = SimpleDense(units=32, activation=tf.nn.relu) #建立一個Dense層
input = tf.ones(shape=(2, 784)) #建立一個輸入張量
output = my_dense(input) #進行正向傳播
print(output.shape)

#自動推論出權重的shape : 即時建構 layer 的權重
#該張量的第一軸維度已經固定成32了，因此後面接續的 layer 的 shape 必須為(批次量, 32)
layer = layers.Dense(32, activation="relu") 

#建構模型
model = models.Sequential([
    layers.Dense(32, activation="relu"),
    layers.Dense(32)
])

#%% 使用 NaiveDense 需要明確的設定各層的輸入大小傳入建構子，不方便使用
class NaiveDense:
    def __init__(self, input_size, output_size, activation):
        keras.activations.get(activation)

        w_shape = (input_size, output_size)
        w_initial_value = tf.random.uniform(w_shape, minval=0, maxval=1e-1)
        self.W = tf.Variable(w_initial_value)

        b_shape = (output_size,)
        b_initial_value = tf.zeros(b_shape)
        self.b = tf.Variable(b_initial_value)

    def __call__(self, inputs):
        return self.activation(tf.matmul(inputs, self.W) + self.b)

    @property
    def weights(self):
        return [self.W, self.b]
model = models.Sequential([
    NaiveDense(input_dim=784, units=32, activation="relu"),
    NaiveDense(input_dim=32, units=64, activation="relu"),
    NaiveDense(input_dim=64, units=32, activation="relu"),
    NaiveDense(input_dim=32, units=10, activation="softmax")
]) 

#%% layer 類別裡的 __call__ 裡的重要作業(像自動呼叫 build) Keras 已經都定義好了，我們只需要自己修改 build 和 call 這兩個方法就好
class SimpleDense(keras.layers.Layer): #繼承Layer類別
    def __init__(self, units, activation=None):
        super().__init__()
        self.units = units 
        keras.activations.get(activation)
    def build(self, input_shape): #建立權重
        input_dim = input_shape[-1] #取得輸入維度大小
        self.W = self.add_weight(shape=(input_dim, self.units), initializer='random_normal') #建立權重W
        self.b = self.add_weight(shape=(self.units,), initializer='zeros') #建立偏移量b
    def call(self, inputs): #定義正向傳播
        y = tf.matmul(inputs, self.W) + self.b #計算線性組合
        if self.activation is not None: #如果有設定激活函數
            y = self.activation(y) #套用激活函數
        return y #回傳輸出
#改用有自動推算 shape 功能的 SimpleDense
model = keras.Sequential([
    SimpleDense(32, activation='relu', input_shape=(784,)), #第一層需要先指定大小才能開始建立 而後續的二三四層則是會依照前面層的大小來建立
    SimpleDense(64, activation='relu'),
    SimpleDense(32, activation='relu'),
    SimpleDense(10, activation='softmax')
]) 

#3-6-3 編譯(Compile)階段 : 定義學習的過程
#%% compile() 參數介紹
model = keras.Sequential([keras.layers.Dense(1)]) #建立一個模型
# compile方法一
model.compile(optimizer='adam', #用名稱(大小寫沒有區別)指定優化器(決定如何根據損失函數更新神經網路使損失值變小)
                loss='mean_squared_error', #用名稱(大小寫有差)指定損失函數
                metrics=['accuracy']) #指定評量指標(給使用者觀察用的) 
# compile方法二
model.compile(optimizer=keras.optimizers.RMSprop(), #用物件指定優化器
                loss=keras.losses.meanSquaredError(), #用物件指定損失函數
                metrics=[keras.metrics.BinaryAccuracy()]) #用物件指定評量指標

#%% 將 learning_rate帶入 optimizer 物件, 更改學習率
#有錯誤正常，因為沒有設定my_custom_loss和my_custom_metric_1,my_custom_metric_2
model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=1e-4), #指定學習率，沒有指定則使用預設值
                loss=my_custom_loss, #傳入自製損失函數 # type: ignore
                metrics=[my_custom_metric_1,my_custom_metric_2])  #傳入自製指標 # type: ignore
#通常不需要自製損失函數和指標，因為Keras已經提供了許多常用的損失函數和指標

#3-6-5 搞懂fit()方法
#%% 建立Dense層
class SimpleDense(keras.layers.Layer): #繼承Layer類別
    def __init__(self, units, activation=None):
        super().__init__()
        self.units = units 
        self.activation = keras.activations.get(activation)  # 確保激活函數是可調用的
    def build(self, input_shape): #建立權重
        input_dim = input_shape[-1] #取得輸入維度大小
        self.W = self.add_weight(shape=(input_dim, self.units), initializer='random_normal') #建立權重W
        self.b = self.add_weight(shape=(self.units,), initializer='zeros') #建立偏移量b
    def call(self, inputs): #定義正向傳播
        y = tf.matmul(inputs, self.W) + self.b #計算線性組合
        if self.activation is not None: #如果有設定激活函數
            y = self.activation(y) #套用激活函數
        return y #回傳輸出

#%% 建立模型
model = keras.Sequential([
    keras.layers.InputLayer(input_shape=(2,)),  # 指定輸入形狀
    SimpleDense(32, activation='relu'),  # 第一層
    SimpleDense(64, activation='relu'),  # 第二層
    SimpleDense(32, activation='relu'),  # 第三層
    SimpleDense(1, activation='sigmoid')  # 輸出層
]) 
#%% 編譯模型(定義學習的方法)
model.compile(optimizer='rmsprop', #用名稱(大小寫沒有區別)指定優化器(決定如何根據損失函數更新神經網路使損失值變小)
                loss='binary_crossentropy', #用名稱(大小寫有差,皆為小寫)指定損失函數
                metrics=['binary_accuracy']) #指定評量指標(給使用者觀察用的) 
#%% 訓練模型
history = model.fit(
    inputs,
    targets,
    epochs=10,
    batch_size=128) #進行10次訓練，每次訓練使用128個樣本
print(history.history) #輸出訓練過程中的損失值和評量指標值

#3-6-6 用驗證資料來監控損失值和指標
#%% 3.24 使用 validation_data 參數
model = keras.Sequential([keras.layers.Dense(1)])
model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=0.1),
              loss=keras.losses.MeanSquaredError(),
              metrics=[keras.metrics.BinaryAccuracy()])

indices_permutation = np.random.permutation(len(inputs)) #打亂資料順序
shuffled_inputs = inputs[indices_permutation] #打亂的資料
shuffled_targets = targets[indices_permutation] #打亂的目標值

num_validation_samples = int(0.3 * len(inputs)) #30%的資料作為驗證資料
val_inputs = shuffled_inputs[:num_validation_samples] #取出驗證資料
val_targets = shuffled_targets[:num_validation_samples] #取出驗證目標值
training_inputs = shuffled_inputs[num_validation_samples:] #取出訓練資料
training_targets = shuffled_targets[num_validation_samples:] #取出訓練目標值
history = model.fit(
    training_inputs,
    training_targets,
    epochs=10,
    batch_size=32,
    validation_data=(val_inputs, val_targets) #使用驗證資料進行驗證
)
print(history.history) #輸出訓練過程中的損失值和評量指標值

#%% 訓練(fit)結束後，可以使用evaluate方法來使用特定的測試資料集來評估模型
loss_and_metrics = model.evaluate(val_inputs, val_targets, batch_size=128)
#模型訓練結束後可以使用新的資料來進行預測，也就是推論(inference)
#%% 使用 所有 新的資料進行推論val_inputs可以替換為其他NumPy陣列或TensorFlow張量並回傳一個TensorFlow張量
predictions = model(val_inputs) 
print(predictions[:10])
#%% 使用 predict 方法來使用新資料的 小部分 進行推論
predictions = model.predict(val_inputs, batch_size=128)
print(predictions[:10])
